### PR TITLE
Allow a stack to be specified in yaml

### PR DESCRIPTION
**Pull request by delfick**
possible solution for https://github.com/delfick/bespin/issues/9

### CODE REVIEW

**Comment by atward** [comment]
The cloudformation YAML has macros. You'll probably have to use a custom filetype. I was using cfn_flip from awslabs.

```
from cfn_flip.custom_yaml import yaml as cfn_yaml
```

**Comment by delfick** [comment]
Ok, I've added cfn_flip to the PR.

Are you able to see that it works?

**Comment by atward** [comment]
Doesn't like the OrderedDict from cfn_flip

Before 
```
13:13:03 INFO    bespin.collector Converting bespin
13:13:03 INFO    bespin.collector Converting environments
13:13:03 INFO    bespin.collector Converting stacks.managedpolicies
13:13:03 INFO    bespin.collector Converting stacks.roles_cloudformation
Layer 0
    Stack infra-iam-managedpolicies
    Stack infra-iam-roles-cloudformation
```

```
> /Users/adamward/git/wha/git/bespin/bespin/collector.py(53)find_missing_config()
-> if "stacks" not in self.configuration:
(Pdb) from pprint import pprint
(Pdb) pprint(self.configuration.__dict__)
{'___getitem___cached': {'passwords': {False: (KeyError(<Path(passwords)>,),
                                               True)},
                         'stacks': {False: (MergedOptions(stacks), False)}},
 '___getitem___expected_version': [1, 0, 0],
 '_values_for_cached': {<Path(stacks)>: {False: '1491362685.38:0.681177214179'},
                        <Path(passwords)>: {False: '1491362685.38:0.582119577744'}},
 '_values_for_expected_version': [1, 0, 0],
 '_values_for_value_cache': {<Path(stacks)>: {False: [({'managedpolicies': {'params_yaml': {},
                                                                            'stack_name': 'infra-iam-managedpolicies',
                                                                            'stack_yaml': '00_iam/00_managedpolicies/template.yaml'},
                                                        'roles_cloudformation': {'params_yaml': {},
                                                                                 'stack_name': 'infra-iam-roles-cloudformation',
                                                                                 'stack_yaml': '00_iam/01_roles-cloudformation/template.yaml'}},
                                                       False)]},
                             <Path(passwords)>: {False: []}},
 'converters': <option_merge.converter.Converters object at 0x10f603b10>,
 'dont_prefix': [<class 'input_algorithms.dictobj.dictobj'>],
 'ignore_converters': False,
 'prefix_list': [],
 'prefix_string': '',
 'storage': <option_merge.storage.Storage object at 0x10f603c50>}
```

After
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Something went wrong! -- BadConfiguration
	"Bad configuration. Didn't find any environments configuration"
Traceback (most recent call last):
  File "/Users/adamward/.virtualenvs/bespin/bin/bespin", line 11, in <module>
    load_entry_point('bespin', 'console_scripts', 'bespin')()
  File "/Users/adamward/.virtualenvs/bespin/lib/python2.7/site-packages/delfick_app.py", line 188, in main
    app.mainline(argv, **execute_args)
  File "/Users/adamward/.virtualenvs/bespin/lib/python2.7/site-packages/delfick_app.py", line 260, in mainline
    self.execute(args_obj, args_dict, extra_args, handler, **execute_args)
  File "/Users/adamward/git/wha/git/bespin/bespin/executor.py", line 27, in execute
    collector.prepare(args_obj.bespin_config.name, args_dict)
  File "/Users/adamward/.virtualenvs/bespin/lib/python2.7/site-packages/option_merge/collector.py", line 175, in prepare
    self.find_missing_config(self.configuration)
  File "/Users/adamward/git/wha/git/bespin/bespin/collector.py", line 54, in find_missing_config
    raise self.BadConfigurationErrorKls("Didn't find any environments configuration")
bespin.errors.BadConfiguration: "Bad configuration. Didn't find any environments configuration"
```

```
> /Users/adamward/git/wha/git/bespin/bespin/collector.py(53)find_missing_config()
-> if "stacks" not in self.configuration:
(Pdb) pprint(self.configuration.__dict__)
{'___getitem___cached': {'stacks': {False: (MergedOptions(stacks), False)}},
 '___getitem___expected_version': [1, 0, 0],
 '_values_for_cached': {<Path(stacks)>: {False: '1491362788.72:0.864735532808'}},
 '_values_for_expected_version': [1, 0, 0],
 '_values_for_value_cache': {<Path(stacks)>: {False: [(OrderedDict([('managedpolicies', OrderedDict([('stack_yaml', '00_iam/00_managedpolicies/template.yaml'), ('stack_name', 'infra-iam-managedpolicies'), ('params_yaml', OrderedDict())])), ('roles_cloudformation', OrderedDict([('stack_yaml', '00_iam/01_roles-cloudformation/template.yaml'), ('stack_name', 'infra-iam-roles-cloudformation'), ('params_yaml', OrderedDict())]))]),
                                                       False)]}},
 'converters': <option_merge.converter.Converters object at 0x10ff87750>,
 'dont_prefix': [<class 'input_algorithms.dictobj.dictobj'>],
 'ignore_converters': False,
 'prefix_list': [],
 'prefix_string': '',
 'storage': <option_merge.storage.Storage object at 0x10ff87890>}
```

**Comment by delfick** [comment]
ah, I see, the OrderedDict breaks the MergedOptions stuff......

and because cfn_flip changes the global pyYaml module it effects everything to be an OrderedDict and so nothing works....

**Comment by atward** [comment]
I'll run with this for a bit. We're going to need a proper CF yaml loader.

```
TypeError: datetime.date(2010, 9, 9) is not JSON serializable
```

eg `AWSTemplateFormatVersion: 2010-09-09` vs `AWSTemplateFormatVersion: "2010-09-09"`

**Comment by atward** [comment]
Overall the biggest problem I have with this approach is that it's converting the YAML template into a JSON template and uploading the result to CF. This will lose all of the documentation, and also introduce a bunch of issues for some of the YAML types (already had issues with datetime, null, int vs strings).

Would it be possible to validate and upload the original template?

**Comment by delfick** [comment]
Yeah, that makes sense, shouldn't be too hard to make it do that

**Comment by delfick** [comment]
ok, it reads in the yaml as just a string and does passes that around now.

File: bespin/option_spec/stack_objs.py [hunk 1]
```diff
@@ -206,8 +207,18 @@ def s3(self):
...
-    def stack_json_obj(self):
-        return self.stack_json
+    def stack_obj(self):
+        if self.stack_json is not NotSpecified:
+            return self.stack_json
+        else:
+            return self.stack_yaml
+
+    @property
+    def dumped_stack_obj(self):
```
Comment (atward) [inline bespin/option_spec/stack_objs.py:217]: Needs to return something

**Review by atward — CHANGES_REQUESTED** [review]
Should also rebase this against master.
Otherwise appears to work (once patched).

File: bespin/option_spec/stack_objs.py (hunk 1 above)
Comment (delfick) [reply bespin/option_spec/stack_objs.py:217]: oops

**Comment by delfick** [comment]
ok, fixed that bug and rebased

**Comment by atward** [comment]
:shipit:

**Comment by delfick** [comment]
0.6.1 released

**Pull request merged**